tests/controller_tests/test_historical_ingest.py

```diff
@@ -0,0 +1,420 @@
+    populate_db_historical_data(r2_workbook_mockup, INGEST_MAPPINGS)
+
+    # test that data has not been duplicated by re-ingest
+    org = Organisation.query.filter_by(organisation_name="United Federation Of Planets").all()
+    assert len(org) == 1
+
+    prog = Programme.query.filter_by(programme_id="TD-FED").all()
+    assert len(prog) == 1
+
+    child = Funding.query.filter_by(funding_source_name="Replicators").all()
+    assert len(child) == 1
+
+
+def test_historical_populate_db_with_round_three(test_client, r2_workbook_mockup, r1_workbook_mockup):
+    """Tests that multiple rounds can be ingested without later rounds being overwritten.
+
+    Test also ensures that re-ingestion of Round 2 does not cause unexpected behaviour.
+    """
+    sub = Submission(
+        submission_id="S-R03-01",
+        submission_date=datetime(2023, 5, 1),
+        reporting_period_start=datetime(2023, 4, 1),
+        reporting_period_end=datetime(2023, 4, 1),
+        reporting_round=3,
+    )
+    db.session.add(sub)
+    read_sub = Submission.query.first()
+    organisation = Organisation(
+        organisation_name="Romulan Star Empire",
+        geography="Romulas",
+    )
+    db.session.add(organisation)
+    read_org = Organisation.query.first()
+    prog = Programme(
+        programme_id="TD-ROM",
+        programme_name="Tal Shiar",
+        fund_type_id="TD",
+        organisation_id=read_org.id,
+    )
+    db.session.add(prog)
+    read_prog = Programme.query.first()
+    proj = Project(
+        project_id="TD-ROM-01",
+        submission_id=read_sub.id,
+        programme_id=read_prog.id,
+        project_name="Patrol the neutral zone",
+        primary_intervention_theme="N/A",
+        location_multiplicity="Single",
+        locations="ROM EMP",
+    )
+    db.session.add(proj)
+
+    populate_db_historical_data(r2_workbook_mockup, INGEST_MAPPINGS)
+
+    # check R3 data is still present
+    org = Organisation.query.filter_by(organisation_name="Romulan Star Empire").first()
+    assert org is not None
+
+    prog = Programme.query.filter_by(programme_id="TD-ROM").first()
+    assert prog is not None
+
+    child = Project.query.filter_by(project_id="TD-ROM-01").first()
+    assert child is not None
+
+    populate_db_historical_data(r1_workbook_mockup, INGEST_MAPPINGS)
+
+    # only expect one result as existing org should not be updated
+    org = Organisation.query.filter_by(organisation_name="Romulan Star Empire").all()
+    assert len(org) == 1
+    assert org[0] == read_org
+
+    # expect R3 programme to not be overwritten
+    prog = Programme.query.filter_by(programme_id="TD-ROM").all()
+    assert len(prog) == 1
+    assert prog[0] == read_prog
+
+    # expect both projects for the programme have been ingested
+    proj1 = Project.query.filter_by(project_id="TD-ROM-01").first()
+    proj2 = Project.query.filter_by(project_id="TD-ROM-02").first()
+    assert proj1 is not None and proj2 is not None
```

Comment:
The only thing we could potentially add to tests is the case where R1 exists, R3 does not, then R2 comes in and does update a programme ref'd by R1, without altering the R1 child project data itself?

Reply:
I've updated the tests to include this scenario, checking to ensure (a) programme is updated, but that (b) programme's child is not deleted/changed.